Improve updating CKAN resources

tomhukins opened the pull request:
Each CKAN data set might have one or more resources.  Previously, the
code disregarded any existing resources and recreated them with a new
ID.  This broke bookmarking resource URLs and following links to
resources if when the shared data harvester runs in between generating
links and requesting them.

This commit retains the IDs, and consequently URLs, of resources, but it
does so by removing them briefly before recreating them with their
previous IDs.  Calling package_patch() instead of package_update() might
help with this, but when I tried this the CKAN client disliked the
StringIO objects.  It's also possible that package_patch() would handle
fields removed from the metadata badly, by leaving them intact instead
of removing them, so we'd need to think carefully about the logic of
this approach.

tomhukins committed:
Improve updating CKAN resources

Each CKAN data set might have one or more resources.  Previously, the
code disregarded any existing resources and recreated them with a new
ID.  This broke bookmarking resource URLs and following links to
resources if when the shared data harvester runs in between generating
links and requesting them.

This commit retains the IDs, and consequently URLs, of resources, but it
does so by removing them briefly before recreating them with their
previous IDs.  Calling package_patch() instead of package_update() might
help with this, but when I tried this the CKAN client disliked the
StringIO objects.  It's also possible that package_patch() would handle
fields removed from the metadata badly, by leaving them intact instead
of removing them, so we'd need to think carefully about the logic of
this approach.

croesus commented on src/python/ib1/openenergy/support/ckan.py:
```diff
@@ -188,18 +188,29 @@ def map_date(item):
+                            existing_resources.pop(res_name)
+                        else:
+                            ckan.action.resource_create(**resource)
+                    for obsolete_resource_name in existing_resources:
```
Is this step necessary if package_update removed existing resources? Or are the previous resources left orphaned?

tomhukins replied:
Good question.  With the current code, I suspect this does nothing, but I haven't checked in detail.  I originally wrote this when I was using `package_patch()` instead of `package_update()`.  I think the best thing to do is for me to write a comment above this code mentioning that it probably does nothing currently (but we should check for orphaned  resources), but if/when we move to `package_patch()` we will need it.

tomhukins committed:
Add a comment explaining some unintuitive code

Chris points out in code review that we probably don't need this given
that package_update() deletes any existing resources.  However, if we
switch to package_patch(), we will need this block, and we might need it
in case orphaned resources exist.  I think they don't, but we should
test any changes thoroughly before removing this code.
